gazebo/Server.cc

```diff
@@ -542,6 +548,110 @@ void Server::Fini()
   gazebo::shutdown();
 }
 
+std::map<std::string, gazebo::common::Time> sensorsLastMeasurementTime;
+std::map<std::string, gazebo::common::Time> worldLastMeasurementTime;
+struct sensorPerformanceMetricsType
+{
+  double sensorRealUpdateRate;
+  double sensorSimUpdateRate;
+  double sensorFPS;
+};
+std::map<std::string, struct sensorPerformanceMetricsType> sensorPerformanceMetrics;
+/// \brief Publisher for run-time simulation performance metrics..
+transport::PublisherPtr performanceMetricsPub;
+
+void PublishPerformanceMetrics()
+{
+  physics::WorldPtr world = physics::get_world("default");
+
+  /// Outgoing run-time simulation performance metrics.
+  msgs::PerformanceMetrics performanceMetricsMsg;
+
+  // Real time factor
+  common::Time realTime = world->RealTime();
+  common::Time simTime = world->SimTime();
+  if (realTime == 0)
+    simTime = 0;
+  else
+    simTime = simTime / realTime;
+
+  if (simTime > 0)
+    performanceMetricsMsg.set_real_time_factor(simTime.Double());
+  else
+    performanceMetricsMsg.set_real_time_factor(0.0);
+
+  /// update sim time for sensors
+  for (auto model: world->Models())
+  {
+    for (auto link: model->GetLinks())
+    {
+      for (unsigned int i = 0; i < link->GetSensorCount(); i++)
+      {
+        std::string name = link->GetSensorName(i);
+        sensors::SensorPtr sensor = sensors::get_sensor(name);
+
+        auto ret = sensorsLastMeasurementTime.insert(
+            std::pair<std::string, gazebo::common::Time>(name, 0));
+        worldLastMeasurementTime.insert(
+                std::pair<std::string, gazebo::common::Time>(name, 0));
+        if (ret.second==false)
+        {
+          double updateSimRate =
+            (sensor->LastMeasurementTime() - sensorsLastMeasurementTime[name]).Double();
+          double updateRealRate =
+            (world->RealTime() - worldLastMeasurementTime[name]).Double();
+          sensorsLastMeasurementTime[name] = sensor->LastMeasurementTime();
```

Comment:
move this down below since we only need to set the var if `updateSimRate > 0.0`?